txscript/opcode.go

```diff
@@ -1918,53 +1907,94 @@ func opcodeCheckMultiSig(op *parsedOpcode, s *Script) error {
 			len(dummy))
 	}
 
-	if len(signatures) == 0 {
-		s.dstack.PushBool(nsig == 0)
-		return nil
-	}
-
 	// Trim OP_CODESEPARATORs
 	script := s.subScript()
 
 	// Remove any of the signatures that happen to be in the script.
 	// can't sign somthing containing the signature you're making, after
 	// all
-	for i := range sigStrings {
-		script = removeOpcodeByData(script, sigStrings[i])
-	}
-
-	curPk := 0
-	for i := range signatures {
-		// check signatures.
-		success := false
-
-		hash := calcScriptHash(script, SigHashType(signatures[i].ht),
-			&s.tx, s.txidx)
-	inner:
-		// Find first pubkey that successfully validates signature.
-		// we start off the search from the key that was successful
-		// last time.
-		for ; curPk < len(pubKeys); curPk++ {
-			if pubKeys[curPk] == nil {
-				pubKeys[curPk], err =
-					btcec.ParsePubKey(pubKeyStrings[curPk],
-						btcec.S256())
-				if err != nil {
-					continue
+	for _, signature := range signatures {
+		script = removeOpcodeByData(script, signature)
+	}
+
+	if len(signatures) != numSignatures {
+		s.dstack.PushBool(false)
+		return nil
+	}
+
+	success := true
+	pubKeyIdx := 0
+
+done:
+	for _, rawSignature := range signatures {
+		if !success {
+			break done
+		}
+
+		// Get the hashtype
+		hashType := SigHashType(rawSignature[len(rawSignature)-1])
+
+		// Trim off the hashtype
+		signature := rawSignature[:len(rawSignature)-1]
+
+		if s.verifyStrictEncoding {
```

Comment:
I'd much rather there be a `checkSignatureEncoding` and `checkPubKeyEncoding` which more closely mirrors BC here (and in `opcodeCheckSig` too) rather than checking the flags in multiple places.  It's much easier to ensure consensus if the same naming and logic is followed.  Also, there is less chance of forgetting to check for a condition in multiple places when it is centralized into the `checkX` functions.  For example, adding the Low S checks will only have to be done in `checkSignatureEncoding`, after checking for the flag,  instead of multiple places.